### ISSUE TITLE
Add support for AspNet Core MVC 1

### ISSUE THREAD

**Issue by kevinkuszyk** [T4MVC/T4MVC#65]
I _think_ the T4MVC T4 template will work with AspNet Core MVC 1, but `T4MVCExtensions` doesn't compile against `coreclr`.

We also need #64 before we can ship this.

**Comment by kevinkuszyk** [comment]
@davidebbo thoughts on a V4 milestone for this and #64?

**Comment by davidebbo** [comment]
Whether it should be v3 depends on whether it entails is a breaking change to .NET 4.x users (based on semver semantics). But hopefully it doesn't, and is separate work that makes Core work without breaking existing.

Do you think it can be the exact same template or will it be a different one?

**Comment by kevinkuszyk** [comment]
Hopefully we can use one xproj / `project.json` project to build `T4MVCExtensions` for both `net40` and `dnxcore50` (or whatever the TFM will be called when RC2 drops).   I didn't get any time to look too much into this yet.

> Do you think it can be the exact same template or will it be a different one?

Hopefully the same one will work (famous last words!).  Unfortunately I can't test in in VS as it just hangs at the moment.  I'll try again when the next tooling update is out.  Do you get internal VS builds you can try it on?

If the template needs some changes to make it work, maybe that effort would be better spent on R4MVC?

**Comment by davidebbo** [comment]
Yes, I saw that hang too when I tried, but I don't know what the deal is. Would be best to open an issue in core with hopefully a simple repro that those guys can look at (I run regular VS).

We can evaluate whether to go with R4MVC based on how drastically different things become in Core.

**Comment by kevinkuszyk** [comment]
I asked @sayedihashimi to re-open [aspnet/Home#272](https://github.com/aspnet/Home/issues/272#issuecomment-180278896) earlier.  Hopefully that's enough to get it on the tooling team's radar.

There's also a sample repro here: https://github.com/kevinkuszyk/aspnet-core-t4mvc.

> We can evaluate whether to go with R4MVC based on how drastically different things become in Core.

Yes, let's make that call once we know how broken it is.

**Comment by davidebbo** [comment]
At the time, I had debugged the template a bit and found that the hang happened in the T4MVC multi-file support. But beyond that I couldn't pinpoint the cause.

**Comment by sayedihashimi** [comment]
See https://github.com/aspnet/Tooling/issues/394

**Comment by kevinkuszyk** [comment]
@davidebbo, based on @mlorbetske suggested fixes in aspnet/Tooling#399, I think there are 3 things we need to do if we want to support MVC Core:
1. Make @mlorbetske's changes to the template so that it can run in both `.csproj` and `.xproj` projects.
2. Generate the correct namespaces if we are in an MVC Core project.
3. Nest the `.generated.cs` files under the `T4MVC.tt` file in `.xproj` projects.
   ![image](https://cloud.githubusercontent.com/assets/2734580/14333593/6f1f501e-fc47-11e5-88d3-d6eef3b3d4d5.png)

I have the T4 template generating in a branch in my [sample repro](https://github.com/kevinkuszyk/aspnet-core-t4mvc/tree/mlorbetske-suggested-fix), but it doesn't build because all the namespaces have changed.

How much effort do you think is involved to make these changes, and do you think I missed anything?

Unfortunately I don't have much time to dedicate to this at the moment (or much knowledge of T4).  Maybe someone in the community can pick this up?

**Comment by davidebbo** [comment]
@kevinkuszyk I also have very limited time to help. But if someone drives a change which doesn't look too scary and makes Core work without breaking current scenarios, I'm all for it.

**Comment by valeriob** [comment]
Hi all, how is t4mvc moving ? What's the status on this based on the recent alpha of msbuild tooling ?

**Comment by kevinkuszyk** [comment]
@valeriob I've not looked at this for a while or had a chance to look at the new MSBuild tooling yet.

If you have the time we'd love to take a PR with this.

We also started a [sister project](https://github.com/T4MVC/R4MVC) to bring similar features to MVC Core, but that project [is also stalled](https://github.com/T4MVC/R4MVC/issues/40) at the moment.

**Comment by valeriob** [comment]
Hi @kevinkuszyk sorry but i do not understand is T4MVC or R4MVC going to work with asp.net core ?
I mean i can't believe that people are actually using asp.net core without strongly typed urls, it's just insane :D

**Comment by kevinkuszyk** [comment]
The ambition is for one or both to support MVC Core.  There are some challenges to get there though:

1. MVC Core is a new framework.  Making a single TT template which supports both is going to be difficult.
2. R4MVC will only support MVC Core, but as it's new we have to start from scratch.

If you have some time to pitch in, maybe you can make it happen!

**Comment by valeriob** [comment]
Yea i'd love to have time for this, and all other projects that we need that are left behind in the transition to core :D

**Comment by valeriob** [comment]
I'm sorry but there are too many things to learn to help with this (project system, nuget, dotNetCore quirks), i'm willing to test the early version but i really think that something that important should be started by who deeply know what we are talking about. I think there is a very big big gap in aspNetCore Mvc right now, i can't understand how are ppl willing to write a web applications with magic strings for creating urls, a real nightmare.

**Comment by artiomchi** [comment]
For the people interested in this, I'm making great progress on this in [T4MVC/R4MVC](https://github.com/T4MVC/R4MVC).

An early alpha build is basically imminent

**Comment by davidebbo** [comment]
@artiomchi awesome, thanks for helping make it happen! Sorry, that I'm not able to get involved in that project, but I fully support the effort in spirit :)
